src/Raven.Server/Documents/Sharding/Operations/ShardedQueryOperation.cs

```diff
@@ -0,0 +1,148 @@
+    {
+        _context = context;
+        _requestHandler = requestHandler;
+        _queryCommands = queryCommands;
+        _includeCompareExchangeValues = includeCompareExchangeValues;
+        _sortingComparer = sortingComparer;
+        ExpectedEtag = expectedEtag;
+    }
+
+    public string ExpectedEtag { get; }
+
+    public HttpRequest HttpRequest { get => _requestHandler.HttpContext.Request; }
+
+    RavenCommand<QueryResult> IShardedOperation<QueryResult, ShardedReadResult<ShardedQueryResult>>.CreateCommandForShard(int shardNumber) => _queryCommands[shardNumber];
+
+    public HashSet<string> MissingIncludes { get; private set; }
+
+    public ShardedQueryResult CombineResults(Memory<QueryResult> results)
+    {
+        var result = new ShardedQueryResult
+        {
+            Results = new List<BlittableJsonReaderObject>()
+        };
+
+        foreach (var cmdResult in results.Span)
+        {
+            result.TotalResults += cmdResult.TotalResults;
+            result.IsStale |= cmdResult.IsStale;
+            result.SkippedResults += cmdResult.SkippedResults;
+            result.IndexName = cmdResult.IndexName;
+            result.IncludedPaths = cmdResult.IncludedPaths;
+            result.ResultEtag = Hashing.Combine(result.ResultEtag, cmdResult.ResultEtag);
+
+            if (result.IndexTimestamp < cmdResult.IndexTimestamp) 
+                result.IndexTimestamp = cmdResult.IndexTimestamp;
+
+            if (result.LastQueryTime < cmdResult.LastQueryTime) 
+                result.LastQueryTime = cmdResult.LastQueryTime;
+
+            if (cmdResult.RaftCommandIndex.HasValue)
+            {
+                if (result.RaftCommandIndex == null || cmdResult.RaftCommandIndex > result.RaftCommandIndex)
+                    result.RaftCommandIndex = cmdResult.RaftCommandIndex;
+            }
+
+            // For includes, we send the includes to all shards, then we merge them together. We do explicitly
+            // support including from another shard, so we'll need to do that again for missing includes
+            // That means also recording the include() call from JS on missing values that we'll need to rerun on
+            // other shards
+
+            if (cmdResult.Includes is { Count: > 0 })
+            {
+                result.Includes ??= new List<BlittableJsonReaderObject>();
+
+                HandleDocumentIncludes(cmdResult, ref result);
+            }
+
+            if (_includeCompareExchangeValues != null && cmdResult.CompareExchangeValueIncludes != null)
+            {
+                HandleCompareExchangeIncludes(cmdResult, ref result);
+            }
+
+            if (_sortingComparer == null)
+            {
+                foreach (BlittableJsonReaderObject item in cmdResult.Results)
+                {
+                    result.Results.Add(item.Clone(_context));
```

Comment:
An optimization, and not sure it is possible... but it would be much nice not to materialize all results here, but return an `IEnumerator<BlittableJsonReaderObject>`

Reply:
I'm not sure if that would be possible. Since the json operation context where those blittables were created on is disposed after the execution of operation:

https://github.com/ravendb/ravendb/blob/f72881e9d2e15e41fc8ecbf2393864eca1961c02/src/Raven.Server/Documents/Sharding/Executors/AbstractExecutor.cs#L82

Reply:
I think it is possible, the same way we do for streaming.
Let's just have a todo here, to see if we can optimize this later.

Reply:
When we handle the streaming we deal with a stream directly so no blittable is involved at the time of calling `Combine(Memory<StreamResult> results)`. We could try to handle the regular querying in streamed fashion as well I think. 
The handling of  queries with sorting could be challenging. 

I will add a TODO

Reply:
Each stream is sorted, so we should be able to sort that on the fly.. but yes a TODO for now :)